### DISCUSSION TITLE
ffmpeg.filter.areverse broken, only short fragments of silence returned, source.duration=inf

### DISCUSSION THREAD

**Discussion by Moonbase59** [savonet/liquidsoap#3745] [User Support]
**Describe the bug**
In preparation of a new silence removal function, I tried to use `ffmpeg.filter.areverse` in an ffmpeg filter chain. When I use it, the script runs very fast, and only extremely short pieces of silence are returned, instead of a song playing backwards.

(The idea behind this, stolen from my "strip-silence" bash script, is to use `ffmpeg.filter.silenceremove` on _both_ ends of a song, so silence before and after can be safely removed, and the song not corrupted like with `blank.eat`.)

Also, after applying the filters, `source.duration` is `inf`.

**To Reproduce**
```
# test-areverse.liq
# 2024-02-12 Moonbase59
#
# ffmpeg.filter.areverse doesn't work

settings.log.level := 3

uri = "~/Musik/Playlists/Radio/Schlager.m3u"
radio = playlist(uri)

# Be sure to have ReplayGain applied before crossing.
radio = amplify(1.,override="replaygain_track_gain",radio)

def reverse(s) =
  def mkfilter(graph) =
    let {audio = audio_track} = source.tracks(s)
    # input
    audio_track = ffmpeg.filter.audio.input(graph, audio_track)
    # reverse audio (end --> start)
    audio_track = ffmpeg.filter.areverse(graph, audio_track)
    # output
    audio_track = ffmpeg.filter.audio.output(graph, audio_track)
    
    # return a filtered source
    source({
      audio = audio_track,
      metadata = track.metadata(audio_track),
      track_marks = track.track_marks(audio_track)
    })
  end
  
  ffmpeg.filter.create(mkfilter)
end

def reverse_wrap(s) =
  s = ffmpeg.raw.encode.audio(%ffmpeg(%audio.raw), s)
  s = reverse(s)
  ffmpeg.raw.decode.audio(s)
end

radio = reverse_wrap(radio)

radio = mksafe(radio)

output.ao(radio)
output.file(%mp3, "test-areverse.mp3", radio)
```

If you comment out the line
```
audio_track = ffmpeg.filter.areverse(graph, audio_track)
```
everything works, so the rest of the code should be okay. I can also use other filters instead of `ffmpeg.filter.areverse` and they work.

In case LS uses the system ffmpeg (4.4.2-0ubuntu0.22.04.1), that should also be okay, since my "strip-silence" bash script works. It uses the same filters.

**Expected behavior**
A continuous stream of songs being played backwards. Real `source.duration` available.

**Version details**

- OS: Linux Mint 21.3
- Version [e.g. `1.3.4`]: Liquidsoap 2.2.4-1

**Install method**
.deb

**Comment by toots** [comment]
I'm not sure that this filter can be used with liquidsoap streams. They are infinite streams!

Please keep in mind that ffmpeg filters operate at the _content_ level. They are not aware of the concept of tracks so, reversing the audio streams means reversing all the audio stream, not just track by track.

**Comment by Moonbase59** [comment]
Yeah, another dead end then. I was hoping I could come up with something that works on a file basis, _before_ they get added to the stream. And of course that should exclude potentially infinite things like remote streams.

I’m currently trying to use an external Python script for that, because I don’t grasp the language good enough to write what’s needed in LS.

**Reply by toots** [reply]
As I was suggesting initially, if you work at the protocol resolution level, you can do whatever you need at a file level before it gets added to the stream.
